homeassistant/components/eight_sleep.py

```diff
@@ -0,0 +1,238 @@
+https://home-assistant.io/components/eight_sleep/
+"""
+import asyncio
+import logging
+import os
+from datetime import timedelta
+
+import voluptuous as vol
+
+from homeassistant.core import callback
+from homeassistant.config import load_yaml_config_file
+from homeassistant.const import (
+    CONF_USERNAME, CONF_PASSWORD, CONF_SENSORS, CONF_BINARY_SENSORS,
+    ATTR_ENTITY_ID, EVENT_HOMEASSISTANT_STOP)
+from homeassistant.helpers import discovery
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_send, async_dispatcher_connect)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_point_in_utc_time
+from homeassistant.util.dt import utcnow
+
+REQUIREMENTS = ['pyeight==0.0.2']
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_PARTNER = 'partner'
+
+DATA_EIGHT = 'eight_sleep'
+DEFAULT_PARTNER = False
+DOMAIN = 'eight_sleep'
+
+HEAT_ENTITY = 'heat'
+USER_ENTITY = 'user'
+
+HEAT_SCAN_INTERVAL = timedelta(seconds=60)
+USER_SCAN_INTERVAL = timedelta(seconds=300)
+
+SIGNAL_UPDATE_HEAT = 'eight_heat_update'
+SIGNAL_UPDATE_USER = 'eight_user_update'
+
+NAME_MAP = {
+    'left_current_sleep': 'Left Sleep Session',
+    'left_last_sleep': 'Left Previous Sleep Session',
+    'left_bed_state': 'Left Bed State',
+    'left_presence': 'Left Bed Presence',
+    'right_current_sleep': 'Right Sleep Session',
+    'right_last_sleep': 'Right Previous Sleep Session',
+    'right_bed_state': 'Right Bed State',
+    'right_presence': 'Right Bed Presence',
+}
+
+SENSORS = ['current_sleep', 'last_sleep', 'bed_state']
+
+SERVICE_HEAT_SET = 'heat_set'
+
+ATTR_TARGET_HEAT = 'target'
+ATTR_HEAT_DURATION = 'duration'
+
+VALID_TARGET_HEAT = vol.All(vol.Coerce(int), vol.Clamp(min=0, max=100))
+VALID_DURATION = vol.All(vol.Coerce(int), vol.Clamp(min=0, max=28800))
+
+SERVICE_EIGHT_SCHEMA = vol.Schema({
+    ATTR_ENTITY_ID: cv.entity_ids,
+    ATTR_TARGET_HEAT: VALID_TARGET_HEAT,
+    ATTR_HEAT_DURATION: VALID_DURATION,
+    })
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_PARTNER, default=DEFAULT_PARTNER): cv.boolean,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Set up the Eight Sleep component."""
+    from pyeight.eight import EightSleep
+
+    conf = config.get(DOMAIN)
+    user = conf.get(CONF_USERNAME)
+    password = conf.get(CONF_PASSWORD)
+    partner = conf.get(CONF_PARTNER)
+
+    if hass.config.time_zone is None:
+        _LOGGER.error('Timezone is not set in Home Assistant.')
+        return False
+
+    timezone = hass.config.time_zone
+
+    eight = EightSleep(user, password, timezone, partner, None, hass.loop)
+
+    hass.data[DATA_EIGHT] = eight
+
+    # Authenticate, build sensors
+    yield from eight.start()
```

Comment:
Would it be possible to validate that the auth is correct here and return `False` if it's incorrect ?